feat: added draggable-image directive

alexfarrugia opened the pull request:
Please see: https://github.com/Mawi137/ngx-image-cropper/issues/494

feat: added allowImagePanning input to enable/disable image panning
feat: added imagePanned event emitter to provide new coordinate changes
refactor(demo): changed package dependencies to angular 13
refactor(demo): utilised behaviour subject for transform object for smoother animation

alexfarrugia commented:
@Mawi137 kind reminder re this pull request.

Added a stackblitz you can play around with here: https://stackblitz.com/edit/image-cropper-kekq3j

**Please note that you have to enable Image panning using the new button in the demo app to try this out**

alexfarrugia commented:
Updated pull request and stackblitz to address two issues: 

1) zooming in without having dragged the image was not updating the image itself (since setTransformCss method was not checking if transform method was truthy or not
2) made panning even smoother by taking the zoom scale into consideration

Mawi137 commented on projects/demo-app/src/app/app.component.ts:
```diff
@@ -17,9 +18,10 @@ export class AppComponent {
   showCropper = false;
   containWithinAspectRatio = false;
-  transform: ImageTransform = {};
+  transform$: BehaviorSubject<ImageTransform> = new BehaviorSubject<ImageTransform>({});
```
Why does this need to be a Subject? It shouldn't be necessary if you always make a copy of the object before changing it.

Mawi137 commented on projects/demo-app/src/app/app.component.ts:
```diff
@@ -28,7 +30,6 @@ export class AppComponent {
 
   imageCropped(event: ImageCroppedEvent) {
     this.croppedImage = event.base64;
-    console.log(event);
```
This console log was left intentionally.

Mawi137 commented on projects/demo-app/src/app/app.component.ts:
```diff
@@ -17,9 +18,10 @@ export class AppComponent {
+  transform$: BehaviorSubject<ImageTransform> = new BehaviorSubject<ImageTransform>({});
   imageURL?: string;
   loading = false;
+  allowPanning$: BehaviorSubject<boolean> = new BehaviorSubject<boolean>(false);
```
A normal boolean should work just fine.

Mawi137 commented on projects/ngx-image-cropper/src/lib/component/image-cropper.component.html:
```diff
@@ -9,6 +9,9 @@
       [style.visibility]="imageVisible ? 'visible' : 'hidden'"
       [style.transform]="safeTransformStyle"
       (load)="imageLoadedInView()"
+      [draggableImage]="allowImagePanning"
```
Instead of a directive, I'd reuse the @HostListeners that are already present. So add the move background image code to the existing `startMove`, ` moveImg` and `moveStop` (can be renamed). 

With css you can set pointer-events to none on the cropper if the `allowImagePanning` is active, I'd also hide the resize squares in that case.

Mawi137 commented on projects/ngx-image-cropper/src/lib/component/image-cropper.component.ts:
```diff
@@ -175,10 +177,16 @@ export class ImageCropperComponent implements OnChanges, OnInit {
 
+  onImageMoved(x: number, y: number): void{
+    if (this.allowImagePanning) {
+            this.imagePanned.next({x, y});
```
Instead I'd update the `this.transform` object and rename the `imagePanned` to `changeTransform` to support 2 way databinding. The image should be cropped as well after the image was moved.

alexfarrugia replied:
hi @Mawi137 , thanks for your feedback, just submitted the code changes. Only point i wasn't so sure about was about removing the pointer events when image panning is allowed. I was thinking it might make sense to still allow the user to move the cropper, what do you think?

I also created an updated stackblitz here if you want to try it out:

https://stackblitz.com/edit/image-cropper-7nypax

Mawi137 reviewed (changes requested):
Hi thanks for the PR. Instead of the directive, I'd try to plugin the background image move into the existing code to reuse as much as possible.

Mawi137 commented:
Hi again, thanks for the changes. I changed a few things and renamed the input to `allowMoveImage`. The feature was release in version 6.2.0!

alexfarrugia commented:
Awesome! Thanks

On Thu, 23 Jun 2022, 21:20 Martijn, ***@***.***> wrote:

> Hi again, thanks for the changes. I changed a few things and renamed the
> input to allowMoveImage. The feature was release in version 6.2.0!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Mawi137/ngx-image-cropper/pull/495#issuecomment-1164780473>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACP6KB5U6TJ3LV4QVKC4VBTVQS2HRANCNFSM5RNQZBEA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
